lib/upgrade/etcd/conversionv1v3/rule.go

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conversionv1v3
+
+import (
+	"fmt"
+	"sync"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	apiv1 "github.com/projectcalico/libcalico-go/lib/apis/v1"
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/numorstring"
+)
+
+// rulesAPIToBackend converts an API Rule structure slice to a Backend Rule structure slice.
+func rulesAPIToBackend(ars []apiv1.Rule) []model.Rule {
+	if ars == nil {
+		return []model.Rule{}
+	}
+
+	brs := make([]model.Rule, len(ars))
+	for idx, ar := range ars {
+		brs[idx] = ruleAPIToBackend(ar)
+	}
+	return brs
+}
+
+// rulesV1BackendToV3API converts a Backend Rule structure slice to an API Rule structure slice.
+func rulesV1BackendToV3API(brs []model.Rule) []apiv3.Rule {
+	if brs == nil {
+		return nil
+	}
+
+	ars := make([]apiv3.Rule, len(brs))
+	for idx, br := range brs {
+		ars[idx] = ruleBackendToAPI(br)
+	}
+	return ars
+}
+
+var logDeprecationOnce sync.Once
+
+// ruleAPIToBackend converts an API Rule structure to a Backend Rule structure.
+func ruleAPIToBackend(ar apiv1.Rule) model.Rule {
+	var icmpCode, icmpType, notICMPCode, notICMPType *int
+	if ar.ICMP != nil {
+		icmpCode = ar.ICMP.Code
+		icmpType = ar.ICMP.Type
+	}
+
+	if ar.NotICMP != nil {
+		notICMPCode = ar.NotICMP.Code
+		notICMPType = ar.NotICMP.Type
+	}
+
+	if ar.Source.Net != nil || ar.Source.NotNet != nil ||
+		ar.Destination.Net != nil || ar.Destination.NotNet != nil {
+		logDeprecationOnce.Do(func() {
```

Comment:
Does this matter to log out? We're going to convert to the right thing anyway, so not really important, right?

Reply:
ok, I'll remove it